src/common/http/server_auth.rs

```diff
@@ -0,0 +1,524 @@
+                    }
+                } else {
+                    Err(ErrorMessage::new(
+                        StatusCode::UNAUTHORIZED,
+                        "No authorization header".to_owned(),
+                    ))
+                }
+            }
+            HttpServerAuthMatcher::Vrl { program } => {
+                let mut target = VrlTarget::new(
+                    Event::Log(LogEvent::from_map(
+                        ObjectMap::from([(
+                            "headers".into(),
+                            Value::Object(
+                                headers
+                                    .iter()
+                                    .map(|(k, v)| {
+                                        (
+                                            KeyString::from(k.to_string()),
+                                            Value::Bytes(Bytes::copy_from_slice(v.as_bytes())),
+                                        )
+                                    })
+                                    .collect::<ObjectMap>(),
+                            ),
+                        )]),
+                        Default::default(),
+                    )),
+                    program.info(),
+                    false,
+                );
+                let timezone = TimeZone::default();
+
+                let result = Runtime::default().resolve(&mut target, program, &timezone);
+                match result.map_err(|e| {
+                    warn!("Handling auth failed: {}", e);
+                    ErrorMessage::new(StatusCode::UNAUTHORIZED, "Auth failed".to_owned())
+                })? {
+                    vrl::core::Value::Boolean(result) => {
+                        if result {
+                            Ok(())
+                        } else {
+                            Err(ErrorMessage::new(
+                                StatusCode::UNAUTHORIZED,
+                                "Auth failed".to_owned(),
+                            ))
+                        }
+                    }
+                    _ => Err(ErrorMessage::new(
+                        StatusCode::UNAUTHORIZED,
+                        "Invalid return value".to_owned(),
+                    )),
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::test_util::random_string;
+    use indoc::indoc;
+
+    use super::*;
+
+    #[test]
+    fn config_should_default_to_basic() {
+        let config = indoc! { r#"
+            username: foo
+            password: bar
+            "#
+        };
+
+        let config: HttpServerAuthConfig = serde_yaml::from_str(config).unwrap();
+
+        assert!(matches!(config, HttpServerAuthConfig::Basic { .. }));
+        if let HttpServerAuthConfig::Basic { username, password } = config {
+            assert_eq!(username, "foo");
+            assert_eq!(password.inner(), "bar");
+        } else {
+            unreachable!();
+        }
+    }
```

Comment:
Nit - This an optional enhancement: 
```suggestion
#[test]
fn config_should_default_to_basic() {
    let config: HttpServerAuthConfig = serde_yaml::from_str(indoc! { r#"
        username: foo
        password: bar
    "# }).unwrap();

    if let HttpServerAuthConfig::Basic { username, password } = config {
        assert_eq!(username, "foo");
        assert_eq!(password.inner(), "bar");
    } else {
        panic!("Expected HttpServerAuthConfig::Basic");
    }
}
```

The above is a bit more concise but I am mostly interested in the:
```
panic!("Expected HttpServerAuthConfig::Basic");
```

Reply:
That makes sense, it will produce a clearer error. I will update.